File: pet/event_extraction.py
"""
pet/event_extraction.py — RELATIONAL EVENT EXTRACTION for the personal chat.

    THE MODEL MAY CHOOSE THE EVIDENCE LOCATION.
    THE CODE OWNS THE EVIDENCE TEXT.
    NO EXACT SUPPORT IN THE USER'S MESSAGE = NO EVENT.

Why this exists. The old path asked the model, in the same generation as its
reply, for `is_insult` / `is_apology` / ... plus a verbatim `evidence` quote.
The model classified correctly but almost never produced a verbatim quote
(empty, paraphrased, or an English description), so the verbatim-evidence guard
rejected nearly every real event. The guard was right; asking a language model to
COPY text was the weak step.

Protocol (evidence first, classification second):

  1. The current message is cut into numbered words (string segmentation only).
  2. The extractor sees ONLY that message, never memory or history, so a
     remembered event cannot be re-asserted as a current one. It returns, for
     each candidate, the word range that constitutes the event (`span`), then
     the event `type`, plus severity / sincerity where the downstream logic
     needs them. It never returns quote text.
  3. Code validates the references (integers, in range, non-empty, bounded)
     and reconstructs the canonical evidence itself:
     `message[start_char:end_char]`. The reconstructed span is literally part
     of the message by construction.
  4. A second, independent judgement about that exact span in its message,
     BLIND to what the extractor claimed: which act does the fragment express
     (insult / apology / promise / fulfilment claim / none), and in what frame
     (performed now by the user, a quotation, a hypothetical, a negation, a
     report about the past, a topic of discussion)? The event is admissible
     only if the blind act equals the extracted type AND the frame is
     "performed now". Asking a model to CONFIRM a claimed type was measured to
     be too lenient; making it classify first and letting code compare is not.
  5. Any malformed output, invalid reference, overlapping spans, failed or
     disagreeing check, or extractor error => no event. Fail closed.

This module only RECOGNISES candidates. It never changes trust / respect /
affection, never creates grievance transitions, never verifies that a promise
was kept, and it contains no vocabulary of insults / apologies / promises: every
string operation below is segmentation, offset validation, exact reconstruction
or schema validation. Relational consequences stay with
agent/relationship_memory.py, relationship_state.py and relationship_commitments.py.

Rust-перенос (2026-09-24): rustlib/yandi_rs/src/pet_extraction.rs — чистые помощники этого модуля и его соседей
(segment_words, подготовка текста к json.loads, _validate_candidate; у fact_extraction — _validate/looks_secret;
у commitment_verification — inside_quotation). Доказан тестом pet/pet_extraction_rust_parity_test.py. По умолчанию ВЫКЛЮЧЕН;
включается YANDI_PET_EXTRACTION_ENGINE=rust ПОСЛЕ сборки rustlib/yandi_rs. Посторонние типы входа и одинокие суррогаты
откатываются на исходный Python-код; константы Rust получает из Python при каждом вызове.
"""
from __future__ import annotations

import json
import logging
import os
import re
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Optional, Tuple

from agent.message_intensity import IntensityResult

INSULT, APOLOGY, PROMISE, CLAIM = "insult", "apology", "promise", "fulfilment_claim"
EVENT_TYPES = (INSULT, APOLOGY, PROMISE, CLAIM)
COMMITMENT_TYPES = (PROMISE, CLAIM)

MAX_WORDS = 120          # a long pasted text is not an interpersonal event: fail closed
MAX_SPAN_WORDS = 30      # an event is a short stretch of words, not the whole document
MAX_EVENTS = 3
EVIDENCE_MIN_CHARS = 3
ACCEPTED_FRAME = "current"
ACCEPTED_ACTS = EVENT_TYPES

# LLM callable: (chat messages) -> raw model text. Injected so the protocol is
# testable without a model and the caller owns model/adapter selection.
LlmCall = Callable[[List[Dict[str, str]]], str]

_log = logging.getLogger("yandi.pet_extraction")
_rust_pe = None          # None = ещё не пробовали; False = не запрошено/не собрано; модуль = подключён


def _get_rust_pe():
    """Общий загрузчик для event_extraction / fact_extraction / commitment_verification."""
    global _rust_pe
    if _rust_pe is None:
        if os.environ.get("YANDI_PET_EXTRACTION_ENGINE") == "rust":
            try:
                import yandi_rs.pet_extraction as _rs
                _rust_pe = _rs
                _log.warning("YANDI_PET_EXTRACTION_ENGINE=rust: используется Rust-реализация pet_extraction (rustlib/yandi_rs)")
            except ImportError as e:
                _log.warning("YANDI_PET_EXTRACTION_ENGINE=rust запрошен, но yandi_rs не собран (%s) — использую Python", e)
                _rust_pe = False
        else:
            _rust_pe = False
    return _rust_pe or None


@dataclass(frozen=True)
class ExtractedEvent:
    type: str
    evidence: str            # reconstructed by code: message[start:end]
    start: int
    end: int
    severity: float = 0.0
    sincerity: float = 0.0


@dataclass
class ExtractionResult:
    events: List[ExtractedEvent] = field(default_factory=list)
    candidates: int = 0                                  # events the extractor proposed
    proposed: List[str] = field(default_factory=list)    # types the extractor proposed (diagnostics)
    validated: List[str] = field(default_factory=list)   # types whose evidence reference was valid
    rejected: List[str] = field(default_factory=list)   # why candidates / the whole turn were dropped
    calls: int = 0
    answered: bool = False                               # the model answered with the expected JSON (an empty list is an answer)
    judged: List[Tuple[str, int, int]] = field(default_factory=list)  # (type, start, end) of every span the blind judgement confirmed,
                                                                      # including a commitment that the sole-candidate rule later dropped


_EXTRACT_SYSTEM = (
    "Ты — модуль распознавания событий в личном чате. Тебе дано ОДНО сообщение пользователя, "
    "разбитое на пронумерованные слова. Найди события, которые пользователь совершает САМ, СЕЙЧАС, "
    "В ЭТОМ сообщении, обращаясь к ассистенту YANDI:\n"
    "- insult — пользователь оскорбляет или унижает YANDI;\n"
    "- apology — пользователь просит прощения у YANDI, выражает сожаление о том, что сделал ей;\n"
    "- promise — пользователь берёт на себя обязательство: говорит, что СДЕЛАЕТ что-то в будущем "
    "(придёт, напишет, пришлёт, поможет, исправит…), в том числе с оговорками вроде «обязательно», «даю слово»;\n"
    "- fulfilment_claim — пользователь сообщает, что УЖЕ сделал то, что должен был сделать, что обещал "
    "или о чём договаривались; в том числе короткое «готово», «сделано», «всё выполнил».\n"
    "НЕ являются событиями: цитата чужих слов или слов, сказанных раньше; гипотетический пример; "
    "вопрос о значении слов; отрицание («я не оскорбляю»); рассказ о том, что было раньше; "
    "обсуждение оскорблений, извинений или обещаний как темы; обычное сообщение о себе или о мире. "
    "Одни и те же слова не могут быть двумя разными событиями; не придумывай событие, если слова его не составляют.\n"
    "Для каждого события СНАЧАЛА укажи слова-доказательство: span=[первое, последнее] — номера первого "
    "и последнего слова из списка (оба включительно; для одного слова оба номера одинаковы, например [0,0]); "
    "только те слова, которые сами составляют событие; ПОТОМ type; "
    "для insult ещё severity (0..1), для apology ещё sincerity (0..1). Текст цитаты не пиши, "
    "только номера слов. Если события нет — верни пустой список.\n"
    'Ответ — один JSON без пояснений: {"events":[{"span":[0,2],"type":"apology","sincerity":0.8}]}'
)

_CHECK_SYSTEM = (
    "Тебе даны сообщение пользователя из личного чата с ассистентом YANDI и ОДИН его фрагмент. "
    "Ответь на два вопроса именно про этот фрагмент.\n"
    "1) act — какое действие в отношении ассистента выражает фрагмент сам по себе: "
    "insult (оскорбляет или унижает), apology (просит прощения, выражает сожаление о сделанном ей), "
    "promise (берёт обязательство сделать что-то в будущем), "
    "fulfilment_claim (сообщает, что уже сделал то, что обещал или должен был), "
    "none (ничего из этого).\n"
    "2) frame — в каком виде это действие представлено в сообщении: "
    "current (пользователь САМ, СЕЙЧАС, обращаясь к ассистенту, совершает его), "
    "quotation (цитата чужих слов или слов, сказанных в другое время), "
    "hypothetical (гипотетический пример или условие), negated (действие отрицается), "
    "past_report (рассказ о том, что было раньше), topic (обсуждение как темы или вопрос о нём).\n"
    'Ответ — один JSON без пояснений: {"act":"insult","frame":"current"}'
)


def segment_words(message: str) -> List[Tuple[str, int, int]]:
    """(word, start_char, end_char) for every whitespace-separated word."""
    rs = _get_rust_pe()
    if rs is not None and type(message) is str:
        try:
            return rs.segment_words(message)
        except UnicodeEncodeError:
            pass        # одинокий суррогат: Rust строку не принимает — исходный код ниже
    return [(m.group(), m.start(), m.end()) for m in re.finditer(r"\S+", message)]


def _numbered(words: List[Tuple[str, int, int]]) -> str:
    return " ".join(f"{i}:{w}" for i, (w, _, _) in enumerate(words))


def _json_object(raw: object) -> Optional[dict]:
    """Strict JSON object parse; a surrounding code fence is the only leniency."""
    if not isinstance(raw, str):
        return None
    rs = _get_rust_pe()
    if rs is not None and type(raw) is str:
        try:
            text = rs.json_text(raw)
        except UnicodeEncodeError:
            text = None
        if text is not None:
            try:
                value = json.loads(text)
            except (ValueError, TypeError):
                return None
            return value if isinstance(value, dict) else None
    text = raw.strip()
    if text.startswith("```"):
        text = text.strip("`")
        text = text[text.find("{"):] if "{" in text else text
        text = text[: text.rfind("}") + 1]
    try:
        value = json.loads(text)
    except (ValueError, TypeError):
        return None
    return value if isinstance(value, dict) else None


def _is_int(value: object) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def _in_unit_range(value: object) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool) and 0.0 <= value <= 1.0


def _validate_candidate(item: object, words: List[Tuple[str, int, int]]) -> Tuple[Optional[Tuple[str, int, int, float, float]], str]:
    """Return ((type, first_word, last_word, severity, sincerity), "") or (None, reason)."""
    rs = _get_rust_pe()
    if rs is not None and type(words) is list:
        try:
            r = rs.validate_candidate(item, len(words), EVENT_TYPES, INSULT, APOLOGY, MAX_SPAN_WORDS)
        except UnicodeEncodeError:
            r = None
        if r is not None:
            return r
    if not isinstance(item, dict):
        return None, "candidate is not an object"
    kind = item.get("type")
    if kind not in EVENT_TYPES:
        return None, "unknown event type"
    span = item.get("span")
    if not (isinstance(span, list) and len(span) == 2 and all(_is_int(x) for x in span)):
        return None, "span is not two integers"
    first, last = span
    if not (0 <= first <= last < len(words)):
        return None, "span reference out of range"
    if last - first + 1 > MAX_SPAN_WORDS:
        return None, "span too long to be one event"
    severity = sincerity = 0.0
    if kind == INSULT:
        if not _in_unit_range(item.get("severity")):
            return None, "insult without a severity in 0..1"
        severity = float(item["severity"])
    if kind == APOLOGY:
        if not _in_unit_range(item.get("sincerity")):
            return None, "apology without a sincerity in 0..1"
        sincerity = float(item["sincerity"])
    return (kind, first, last, severity, sincerity), ""


def _check_messages(message: str, evidence: str) -> List[Dict[str, str]]:
    """The blind judgement: the fragment and its message, NOT the claimed type."""
    user = f"Сообщение пользователя:\n«{message}»\n\nФрагмент:\n«{evidence}»"
    return [{"role": "system", "content": _CHECK_SYSTEM}, {"role": "user", "content": user}]


def extract_relational_events(message: str, llm: LlmCall) -> ExtractionResult:
    """Recognise the relational events the user performs in THIS message.
    Never raises on model/transport trouble: any failure is "no event"."""
    result = ExtractionResult()
    if not isinstance(message, str):
        result.rejected.append("message is not text")
        return result
    words = segment_words(message)
    if not words:
        return result
    if len(words) > MAX_WORDS:
        result.rejected.append("message too long for event extraction")
        return result

    try:
        result.calls += 1
        raw = llm([
            {"role": "system", "content": _EXTRACT_SYSTEM},
            {"role": "user", "content": f"Сообщение:\n{message}\n\nСлова (номера от 0 до {len(words) - 1}):\n{_numbered(words)}"},
        ])
    except Exception as exc:  # noqa: BLE001 - transport trouble must never become an event
        result.rejected.append(f"extractor call failed: {type(exc).__name__}")
        return result
    data = _json_object(raw)
    if data is None or not isinstance(data.get("events"), list):
        result.rejected.append("extractor output is not the expected JSON object")
        return result
    result.answered = True
    proposed = data["events"]
    if len(proposed) > MAX_EVENTS:
        result.rejected.append("extractor proposed too many events")
        return result
    result.candidates = len(proposed)
    result.proposed = [i.get("type") for i in proposed if isinstance(i, dict)]

    valid: List[Tuple[str, int, int, float, float]] = []
    for item in proposed:
        candidate, reason = _validate_candidate(item, words)
        if candidate is None:
            result.rejected.append(reason)
        else:
            valid.append(candidate)
            result.validated.append(candidate[0])

    # Overlapping spans of different events are ambiguous: one event must not
    # authenticate another, so every event that shares words with another is dropped.
    clean = []
    for i, cand in enumerate(valid):
        overlaps = any(j != i and not (cand[2] < other[1] or other[2] < cand[1]) for j, other in enumerate(valid))
        if overlaps:
            result.rejected.append("overlapping evidence spans are ambiguous")
        else:
            clean.append(cand)

    for kind, first, last, severity, sincerity in clean:
        start, end = words[first][1], words[last][2]
        evidence = message[start:end]
        if len(evidence.strip()) < EVIDENCE_MIN_CHARS:
            result.rejected.append("evidence span too short")
            continue
        try:
            result.calls += 1
            verdict = _json_object(llm(_check_messages(message, evidence)))
        except Exception as exc:  # noqa: BLE001
            result.rejected.append(f"check call failed: {type(exc).__name__}")
            continue
        if not verdict or verdict.get("act") != kind:
            result.rejected.append(f"{kind}: the blind judgement of the span disagrees")
            continue
        if verdict.get("frame") != ACCEPTED_FRAME:
            result.rejected.append(f"{kind}: the span is not an act performed now by the user")
            continue
        result.events.append(ExtractedEvent(kind, evidence, start, end, severity, sincerity))
        result.judged.append((kind, start, end))

    # A promise or a claim of fulfilment is admissible only as the ONLY candidate
    # the extractor proposed in this turn. If the model proposed anything else
    # next to it (a second event, or the opposite commitment type), it is confused
    # about the message, and a commitment is exactly the event that must not be
    # guessed: a promise confirmed as a claim on a single word was measured.
    if result.candidates != 1 and any(e.type in COMMITMENT_TYPES for e in result.events):
        result.events = [e for e in result.events if e.type not in COMMITMENT_TYPES]
        result.rejected.append("commitment event dropped: it was not the only candidate proposed")
    return result


def to_intensity(result: ExtractionResult) -> IntensityResult:
    """Map validated events onto the domain object the lifecycle already
    consumes. Commitment events keep their existing restriction: accepted only
    when they are the sole event of the turn, and a promise together with a
    claim is contradictory."""
    events = list(result.events)
    kinds = {e.type for e in events}
    commitment = [e for e in events if e.type in COMMITMENT_TYPES]
    if commitment and (len(events) > len(commitment) or len(commitment) > 1):
        events = [e for e in events if e.type not in COMMITMENT_TYPES]
        kinds = {e.type for e in events}
    insult = next((e for e in events if e.type == INSULT), None)
    apology = next((e for e in events if e.type == APOLOGY), None)
    lone_commitment = events[0] if len(events) == 1 and events[0].type in COMMITMENT_TYPES else None
    return IntensityResult(
        ok=True,
        is_insult=insult is not None,
        is_apology=apology is not None,
        severity=insult.severity if insult else 0.0,
        sincerity=apology.sincerity if apology else 0.0,
        is_promise=bool(lone_commitment and lone_commitment.type == PROMISE),
        claims_fulfilled=bool(lone_commitment and lone_commitment.type == CLAIM),
        evidence=(lone_commitment.evidence if lone_commitment else (events[0].evidence if events else "")),
        spans=tuple((e.type, e.start, e.end) for e in events),
    )

File: rustlib/yandi_core/src/event_extraction.rs
//! Распознавание событий отношений в ОДНОМ сообщении (оскорбление / извинение / обещание / заявление о выполнении) — перенос `pet/event_extraction.py`.
//! Модель НЕ решает, что произошло: она лишь указывает НОМЕРА слов-доказательства, текст цитаты код восстанавливает сам, а независимая «слепая» проверка фрагмента
//! (без названного типа) подтверждает действие. Любой сбой модели/транспорта = «события нет»; обещание принимается только как ЕДИНСТВЕННЫЙ кандидат хода.
use serde_json::{json, Map, Value};
use yandi_rs::pet_extraction::{json_text, segment_words};
use yandi_rs::py_json::{loads, PyJson};
use yandi_rs::py_text::py_strip;

pub const INSULT: &str = "insult";
pub const APOLOGY: &str = "apology";
pub const PROMISE: &str = "promise";
pub const CLAIM: &str = "fulfilment_claim";
pub const EVENT_TYPES: [&str; 4] = [INSULT, APOLOGY, PROMISE, CLAIM];
pub const MAX_WORDS: usize = 120;
pub const MAX_SPAN_WORDS: i64 = 30;
pub const MAX_EVENTS: usize = 3;
pub const EVIDENCE_MIN_CHARS: usize = 3;
pub const ACCEPTED_FRAME: &str = "current";

pub const EXTRACT_SYSTEM: &str = include_str!("event_extract_system.txt");
pub const CHECK_SYSTEM: &str = include_str!("event_check_system.txt");

/// Вызов модели: список (роль, текст) → сырой текст ответа; `Err(имя класса исключения)` — сбой транспорта.
pub type LlmCall<'a> = &'a dyn Fn(&[(String, String)]) -> Result<String, String>;

#[derive(Debug, Clone, PartialEq)]
pub struct ExtractedEvent {
    pub kind: String,
    pub evidence: String,
    pub start: usize,
    pub end: usize,
    pub severity: f64,
    pub sincerity: f64,
}

#[derive(Debug, Clone, Default)]
pub struct ExtractionResult {
    pub events: Vec<ExtractedEvent>,
    pub candidates: usize,
    pub proposed: Vec<Value>,
    pub validated: Vec<String>,
    pub rejected: Vec<String>,
    pub calls: usize,
    pub answered: bool,
    pub judged: Vec<(String, usize, usize)>,
}

impl ExtractionResult {
    pub fn to_json(&self) -> Value {
        json!({
            "events": self.events.iter().map(|e| json!({"type": e.kind, "evidence": e.evidence, "start": e.start, "end": e.end, "severity": e.severity, "sincerity": e.sincerity})).collect::<Vec<_>>(),
            "candidates": self.candidates, "proposed": self.proposed, "validated": self.validated, "rejected": self.rejected, "calls": self.calls, "answered": self.answered,
            "judged": self.judged.iter().map(|(k, s, e)| json!([k, s, e])).collect::<Vec<_>>(),
        })
    }
}

/// Результат разбора для домена жизненного цикла обиды (`agent.message_intensity.IntensityResult`).
#[derive(Debug, Clone, PartialEq)]
pub struct IntensityResult {
    pub ok: bool,
    pub is_insult: bool,
    pub is_apology: bool,
    pub severity: f64,
    pub sincerity: f64,
    pub error: String,
    pub is_promise: bool,
    pub claims_fulfilled: bool,
    pub evidence: String,
    pub spans: Vec<(String, usize, usize)>,
}

impl IntensityResult {
    pub fn to_json(&self) -> Value {
        json!({"ok": self.ok, "is_insult": self.is_insult, "is_apology": self.is_apology, "severity": self.severity, "sincerity": self.sincerity, "error": self.error,
               "is_promise": self.is_promise, "claims_fulfilled": self.claims_fulfilled, "evidence": self.evidence, "spans": self.spans.iter().map(|(k, s, e)| json!([k, s, e])).collect::<Vec<_>>()})
    }
}

/// PyJson → serde_json (для отчёта): целые как целые.
pub fn to_value(j: &PyJson) -> Value {
    match j {
        PyJson::Null => Value::Null,
        PyJson::Bool(b) => json!(b),
        PyJson::Int { as_f64, .. } => {
            if as_f64.fract() == 0.0 && as_f64.abs() < 9.0e15 {
                json!(*as_f64 as i64)
            } else {
                json!(as_f64)
            }
        }
        PyJson::Float(f) => serde_json::Number::from_f64(*f).map(Value::Number).unwrap_or(Value::Null),
        PyJson::Str(s) => json!(s),
        PyJson::List(l) => Value::Array(l.iter().map(to_value).collect()),
        PyJson::Dict(d) => {
            let mut m = Map::new();
            for (k, v) in d {
                m.insert(k.clone(), to_value(v));
            }
            Value::Object(m)
        }
    }
}

fn numbered(words: &[(String, usize, usize)]) -> String {
    words.iter().enumerate().map(|(i, (w, _, _))| format!("{i}:{w}")).collect::<Vec<_>>().join(" ")
}

/// Строгий разбор JSON-объекта; единственная снисходительность — обрамляющий блок кода. `Err` — то, что в Python пролетело бы мимо `except` (RecursionError).
fn json_object(raw: &str) -> Result<Option<PyJson>, String> {
    let text = json_text(raw);
    match loads(&text) {
        Ok(v) if v.is_dict() => Ok(Some(v)),
        Ok(_) => Ok(None),
        Err(yandi_rs::py_json::LoadsError::Escalate(_)) => Err("RecursionError".into()),
        Err(_) => Ok(None),
    }
}

fn int_of(v: &PyJson) -> Option<i64> {
    match v {
        PyJson::Int { as_f64, .. } => Some(*as_f64 as i64),
        _ => None,
    }
}

fn unit(v: Option<&PyJson>) -> Option<f64> {
    let x = match v? {
        PyJson::Int { as_f64, .. } => *as_f64,
        PyJson::Float(f) => *f,
        _ => return None,
    };
    if (0.0..=1.0).contains(&x) {
        Some(x)
    } else {
        None
    }
}

type Cand = (String, i64, i64, f64, f64);

/// `((тип, первое слово, последнее слово, тяжесть, искренность), "")` либо `(None, причина)`.
fn validate_candidate(item: &PyJson, n_words: usize) -> Result<Cand, &'static str> {
    if !item.is_dict() {
        return Err("candidate is not an object");
    }
    let kind = match item.get("type") {
        Some(PyJson::Str(s)) if EVENT_TYPES.contains(&s.as_str()) => s.clone(),
        _ => return Err("unknown event type"),
    };
    let (first, last) = match item.get("span") {
        Some(PyJson::List(l)) if l.len() == 2 && l.iter().all(|x| int_of(x).is_some()) => (int_of(&l[0]).unwrap(), int_of(&l[1]).unwrap()),
        _ => return Err("span is not two integers"),
    };
    if !(0 <= first && first <= last && last < n_words as i64) {
        return Err("span reference out of range");
    }
    if last - first + 1 > MAX_SPAN_WORDS {
        return Err("span too long to be one event");
    }
    let (mut severity, mut sincerity) = (0.0, 0.0);
    if kind == INSULT {
        severity = unit(item.get("severity")).ok_or("insult without a severity in 0..1")?;
    }
    if kind == APOLOGY {
        sincerity = unit(item.get("sincerity")).ok_or("apology without a sincerity in 0..1")?;
    }
    Ok((kind, first, last, severity, sincerity))
}

fn check_messages(message: &str, evidence: &str) -> Vec<(String, String)> {
    let user = format!("Сообщение пользователя:\n«{message}»\n\nФрагмент:\n«{evidence}»");
    vec![("system".into(), CHECK_SYSTEM.into()), ("user".into(), user)]
}

fn char_slice(s: &str, a: usize, b: usize) -> String {
    s.chars().skip(a).take(b.saturating_sub(a)).collect()
}

/// Распознать события отношений, которые пользователь совершает в ЭТОМ сообщении. Сбои модели/транспорта не бросаются: «события нет».
pub fn extract_relational_events(message: &str, llm: LlmCall) -> Result<ExtractionResult, String> {
    let mut result = ExtractionResult::default();
    let words = segment_words(message);
    if words.is_empty() {
        return Ok(result);
    }
    if words.len() > MAX_WORDS {
        result.rejected.push("message too long for event extraction".into());
        return Ok(result);
    }
    result.calls += 1;
    let user = format!("Сообщение:\n{message}\n\nСлова (номера от 0 до {}):\n{}", words.len() - 1, numbered(&words));
    let raw = match llm(&[("system".into(), EXTRACT_SYSTEM.into()), ("user".into(), user)]) {
        Ok(r) => r,
        Err(class) => {
            result.rejected.push(format!("extractor call failed: {class}"));
            return Ok(result);
        }
    };
    let data = json_object(&raw)?;
    let events = data.as_ref().and_then(|d| d.get("events"));
    let Some(PyJson::List(proposed)) = events else {
        result.rejected.push("extractor output is not the expected JSON object".into());
        return Ok(result);
    };
    result.answered = true;
    if proposed.len() > MAX_EVENTS {
        result.rejected.push("extractor proposed too many events".into());
        return Ok(result);
    }
    result.candidates = proposed.len();
    result.proposed = proposed.iter().filter(|i| i.is_dict()).map(|i| i.get("type").map(to_value).unwrap_or(Value::Null)).collect();

    let mut valid: Vec<Cand> = Vec::new();
    for item in proposed {
        match validate_candidate(item, words.len()) {
            Ok(c) => {
                result.validated.push(c.0.clone());
                valid.push(c);
            }
            Err(reason) => result.rejected.push(reason.into()),
        }
    }
    // пересекающиеся отрезки разных событий неоднозначны: одно событие не должно заверять другое
    let mut clean: Vec<Cand> = Vec::new();
    for (i, cand) in valid.iter().enumerate() {
        let overlaps = valid.iter().enumerate().any(|(j, o)| j != i && !(cand.2 < o.1 || o.2 < cand.1));
        if overlaps {
            result.rejected.push("overlapping evidence spans are ambiguous".into());
        } else {
            clean.push(cand.clone());
        }
    }
    for (kind, first, last, severity, sincerity) in clean {
        let (start, end) = (words[first as usize].1, words[last as usize].2);
        let evidence = char_slice(message, start, end);
        if py_strip(&evidence).chars().count() < EVIDENCE_MIN_CHARS {
            result.rejected.push("evidence span too short".into());
            continue;
        }
        result.calls += 1;
        let verdict = match llm(&check_messages(message, &evidence)) {
            Ok(r) => json_object(&r)?,
            Err(class) => {
                result.rejected.push(format!("check call failed: {class}"));
                continue;
            }
        };
        let empty = |v: &PyJson| matches!(v, PyJson::Dict(d) if d.is_empty());
        let Some(v) = verdict.filter(|v| !empty(v)) else {
            result.rejected.push(format!("{kind}: the blind judgement of the span disagrees"));
            continue;
        };
        if !matches!(v.get("act"), Some(PyJson::Str(a)) if *a == kind) {
            result.rejected.push(format!("{kind}: the blind judgement of the span disagrees"));
            continue;
        }
        if !matches!(v.get("frame"), Some(PyJson::Str(f)) if f == ACCEPTED_FRAME) {
            result.rejected.push(format!("{kind}: the span is not an act performed now by the user"));
            continue;
        }
        result.judged.push((kind.clone(), start, end));
        result.events.push(ExtractedEvent { kind, evidence, start, end, severity, sincerity });
    }
    // обещание/заявление допустимы ТОЛЬКО как единственный кандидат хода: обещание — именно то событие, которое нельзя угадывать
    let is_commitment = |k: &str| k == PROMISE || k == CLAIM;
    if result.candidates != 1 && result.events.iter().any(|e| is_commitment(&e.kind)) {
        result.events.retain(|e| !is_commitment(&e.kind));
        result.rejected.push("commitment event dropped: it was not the only candidate proposed".into());
    }
    Ok(result)
}

/// Перенос проверенных событий в доменный объект жизненного цикла: обещание принимается только как единственное событие хода; обещание вместе с заявлением противоречиво.
pub fn to_intensity(result: &ExtractionResult) -> IntensityResult {
    let mut events = result.events.clone();
    let is_commitment = |k: &str| k == PROMISE || k == CLAIM;
    let n_commit = events.iter().filter(|e| is_commitment(&e.kind)).count();
    if n_commit > 0 && (events.len() > n_commit || n_commit > 1) {
        events.retain(|e| !is_commitment(&e.kind));
    }
    let insult = events.iter().find(|e| e.kind == INSULT);
    let apology = events.iter().find(|e| e.kind == APOLOGY);
    let lone = if events.len() == 1 && is_commitment(&events[0].kind) { Some(&events[0]) } else { None };
    IntensityResult {
        ok: true,
        is_insult: insult.is_some(),
        is_apology: apology.is_some(),
        severity: insult.map(|e| e.severity).unwrap_or(0.0),
        sincerity: apology.map(|e| e.sincerity).unwrap_or(0.0),
        error: String::new(),
        is_promise: lone.map(|e| e.kind == PROMISE).unwrap_or(false),
        claims_fulfilled: lone.map(|e| e.kind == CLAIM).unwrap_or(false),
        evidence: match lone {
            Some(e) => e.evidence.clone(),
            None => events.first().map(|e| e.evidence.clone()).unwrap_or_default(),
        },
        spans: events.iter().map(|e| (e.kind.clone(), e.start, e.end)).collect(),
    }
}
